File: C#/JlgCommon/Extensions/StringExtensions.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text;
using System.Text.RegularExpressions;

namespace JlgCommon.Extensions
{
    public static class StringExtensions
    {
        public static string ToSqlValidTableOrColumnName(this string str)
        {
            while (str.IndexOf("  ") >= 0)
            {
                str = str.Replace("  ", " ");
            }
            
            var modifiedStr = Regex.Replace(str.Trim(), "[^a-zA-Z0-9_.]+", "_", RegexOptions.Compiled);
            if (modifiedStr.Length > 110)
            {
                modifiedStr = modifiedStr.Substring(0, 110);
            }
            return modifiedStr.ToLower();
        }

        public static bool IsEmail(this string value)
        {
            var validEmailPattern = @"^(?!\.)(""([^""\r\\]|\\[""\r\\])*""|"
            + @"([-a-z0-9!#$%&'*+/=?^_`{|}~]|(?<!\.)\.)*)(?<!\.)"
            + @"@[a-z0-9][\w\.-]*[a-z0-9]\.[a-z][a-z\.]*[a-z]$";

            var regexPattern = new Regex(validEmailPattern, RegexOptions.IgnoreCase);

            return regexPattern.IsMatch(value);
        }

        public static string LowerCaseAndIgnoreSpaces(this string value)
        {
            var comparableString = value.Replace(" ", string.Empty);
            return comparableString.ToLower();
        }

        public static string Capitalize(this string str)
        {
            if (string.IsNullOrEmpty(str)
                || string.IsNullOrEmpty(str.Trim()))
            {
                return str;
            }

            var specialCharacters = new List<string>() { "/", "\\", ".", ";", "," };
            str = str.ToLower().Trim();
            foreach (var specialCharacter in specialCharacters)
            {
                str = str.Replace(specialCharacter, string.Format("{0} ", specialCharacter));
            }

            var words = str.Split(' ');
            var capitalezedPhrase = new StringBuilder();

            foreach (var word in words)
            {
                if (word.Length == 0)
                {
                    continue;
                }

                var wordForEdit = new StringBuilder(word);
                var firstLeterCapitalized = wordForEdit[0].ToString().ToUpper();
                wordForEdit.Remove(0, 1);
                wordForEdit.Insert(0, firstLeterCapitalized);

                capitalezedPhrase.Append(wordForEdit);
                capitalezedPhrase.Append(" ");
            }

            capitalezedPhrase.Remove(capitalezedPhrase.Length - 1, 1);

            var capitalized = capitalezedPhrase.ToString();

            foreach (var specialCharacter in specialCharacters)
            {
                capitalized = capitalized.Replace(string.Format("{0} ", specialCharacter), specialCharacter);
            }

            while (capitalized.Contains("  "))
            {
                capitalized = capitalized.Replace("  ", " ");
            }

            return capitalized;
        }             

        private static Dictionary<Type, TypeConverter> _typeConverters = new Dictionary<Type, TypeConverter>();

        public static T ParseEntry<T>(string entry)
        {
            TypeConverter conv;
            var type = typeof(T);
            if (_typeConverters.ContainsKey(type))
            {
                conv = _typeConverters[type];
            }
            else
            {
                conv = TypeDescriptor.GetConverter(type);
                _typeConverters.Add(type, conv);
            }

            return (T)conv.ConvertFromString(entry);
        }
        
        public static string Stringify<T>(List<T> list)
        {
            var sb = new StringBuilder();

            if (list.Count > 0)
            {
                sb.Append(",");
                for (int i = 0; i < list.Count; i++)
                {
                    sb.Append(list[i]);
                    sb.Append(",");
                }
            }
            return sb.ToString();
        }

        public static List<T> StringifyRevert<T>(string stringigiedList)
        {
            var listElements = stringigiedList.Split(',');
            var list = new List<T>();
            foreach (var element in listElements)
            {
                if (!string.IsNullOrEmpty(element))
                {
                    list.Add(ParseEntry<T>(element));
                }
            }

            return list;
        }
    }
}

File: Java/JlgCommon/src/extensions/StringExtensions.java
package extensions;

import domain.Constants;

import java.util.Arrays;
import java.util.List;

public final class StringExtensions {

    public static String toSqlValidTableOrColumnName(String str)
    {
        str = replaceMultipleSpacesWithSingleSpace(str);
        str = str.trim();
        str = str.replaceAll("[^a-zA-Z0-9_.]+", "_");
        if (str.length() > 110) {
            str = str.substring(0, 110);
        }
        return str.toLowerCase();
    }

    public static boolean isEmail(String email) {
        String validEmailPattern = "^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*@"
                + "[A-Za-z0-9-]+(\\.[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$";
        return email.matches(validEmailPattern);
    }

    public static String lowerCaseAndIgnoreSpaces(String value) {
        String comparableString = value.replaceAll(Constants.SPACE_STRING, Constants.EMPTY_STRING);
        return comparableString.toLowerCase();
    }

    public static String replaceMultipleSpacesWithSingleSpace(String value)
    {
        while (value.contains("  ")) {
            value = value.replaceAll("  ", Constants.SPACE_STRING);
        }
        return value;
    }

    public static String capitalize(String str) {
        if (str.isEmpty()) {
            return str;
        }

        List<String> specialCharacters = Arrays.asList("/", "\\", ".", ";", ",");
        str = str.toLowerCase().trim();
        for (String specialCharacter : specialCharacters) {
            str = str.replace(specialCharacter, String.format("%s ", specialCharacter));
        }

        String[] words = str.split(Constants.SPACE_STRING);
        StringBuilder capitalezedPhrase = new StringBuilder();

        for (String word : words) {
            if (word.length() == 0) {
                continue;
            }
            StringBuilder wordForEdit = new StringBuilder(word);
            String firstLeterCapitalized = ((Character)wordForEdit.charAt(0)).toString().toUpperCase();

            wordForEdit = wordForEdit.delete(0, 1);
            wordForEdit = wordForEdit.insert(0, firstLeterCapitalized);

            capitalezedPhrase.append(wordForEdit);
            capitalezedPhrase.append(Constants.SPACE_STRING);
        }
        capitalezedPhrase = capitalezedPhrase.delete(capitalezedPhrase.length() - 1, capitalezedPhrase.length());
        String capitalized = capitalezedPhrase.toString();
        for (String specialCharacter : specialCharacters) {
            capitalized = capitalized.replace(String.format("%s ", specialCharacter), specialCharacter);
        }
        capitalized = replaceMultipleSpacesWithSingleSpace(capitalized);

        return capitalized;
    }

    private StringExtensions(){
    }
}
